### DISCUSSION TITLE
Hop over gaps

### DISCUSSION THREAD

**Discussion by sanandnarayan** [Jeiwan/uniswapv3-book#41] [Milestone 3. Cross-tick Swaps]
I am curious, why is it mentioned that "it’s not possible to hop over a gap–a swap would be completed partially". Lets say that I want to swap some USDC to eth and I am okay to take a lot of slippage.
I finished half the swap at a range that ends at 5001, and the next range begins at 5003. There is a small gap in between. Would I be able to trade against the next range, and jump over the hop?

**Comment by Jeiwan** [comment]
No. I guess it boils down to the golden rule: liquidity is required to make trades. If there's no liquidity, trading is not possible. If no one sells at 5001 and 5002, how can you buy at these prices?

In case of Uniswap V3, I guess we can say that the transition from 5001 to 5003 is path dependent: there can be different amounts of liquidity between 5001 and 5002, and between 5002 and 5003. If there's a lot of liquidity between 5001 and 5002, you won't even reach 5003, because the liquidity will be enough to fulfill your swap. So, in this sense, skipping gaps breaks paths.

**Comment by sanandnarayan** [comment]
thank you for the answer jeiwan
So here is what I did.

I created a small gap in liquidity
<img width="754" alt="Screen Shot 2022-12-22 at 1 14 56 PM" src="https://user-images.githubusercontent.com/433334/209227128-93adbac7-2d21-43ac-a810-3117ee610f44.png">

And then I commented out this line
<img width="911" alt="Screen Shot 2022-12-22 at 1 15 23 PM" src="https://user-images.githubusercontent.com/433334/209227218-8ef06adc-2ae3-42d4-a8b5-22677b259415.png">

The tests pass (with small modifications to quantity expected and sqrtPriceX96).

The point I am trying to convey is, the user is willing to swap at any price, and liquidity exists for it in the pool, but not consecutive. It is in islands. Then technically we should still be able to swap because both parties (user and lp ers) want to trade at these prices.

We dont have to trade in the gaps, but can hop and trade right?
We can find other ways to stop the while loop (i.e no liquidity). We can have a limit of 2x tick or really far limit prices

**Comment by sanandnarayan** [comment]
I also initialized a new pool on goerli to check if uniswap can hop. I think it can.
Here is a swap that hopped between gaps in liquidity positions.
https://goerli.etherscan.io/tx/0x62bdc51c94f53e7269700ab65059229be18ee1225aa61e3b3acbe7ff56fc559d
